pkg/solidity-utils/contracts/helpers/ScalingHelpers.sol

```diff
@@ -82,7 +82,7 @@ library ScalingHelpers {
         uint256 tokenRate
     ) internal pure returns (uint256) {
         // Do division last, and round scalingFactor * tokenRate up to divide by a larger number.
-        return FixedPoint.divDown(amount, scalingFactor.mulUp(tokenRate));
+        return amount == 0 ? 0 : FixedPoint.divDown(amount, scalingFactor.mulUp(tokenRate));
```

Comment:
Just a triple check for `undo` functions:
- scaling factors depend on token decimals, and can never be 0.
- if token rate is 0, the previous code would revert with panic, but this one would not.

If the rate is 0, then the scaled balances would be 0 anyways. But in that case scaling back doesn't make sense; it's probably better to revert.

While tokens with rate 0 don't make sense per-se, we can't discard some weird intermediate state inside a tx making the rate 0 for tokens. That case should probably revert; continuing the execution might bring unintended consequences.

I'd remove this, and keep the code in `_addLiquidity` and `_removeLiquidity` as it is now without the extra `if`. If the amounts are 0, then let's just let the usual math do the job of multiplying 0 by the rate.